scipy/cluster/hierarchy.py

```diff
@@ -3004,13 +3004,70 @@ def set_link_color_palette(palette):
     _link_line_colors = palette
 
 
+def reorder_leaves(Z, leaves_order):
+    # first, we build a tree
+    n = Z.shape[0] + 1
+    ch = np.zeros((n - 1, 2))
+    for i in range(n - 1):
+        ch[i] = (Z[i][0], Z[i][1])  # ch[i] = children of node n+i
+    Z = Z.astype('int32')
```

Comment:
I believe this is the culprit here. Casting to int gets rid of float valued metrics.

Reply:
Thanks @ihowell! I'm going to improve the complexity from O(n^2) to O(n), and this may lead to other algorithm. I'll keep that bug in mind. Hoping to produce new code in a week

Reply:
@ihowell improved to linear and fixed this bug with casting in https://github.com/scipy/scipy/pull/12049/commits/00954accf853d0ae92e3cbfb2e0d2abbf052a994 ! What do You think about moving this to linkage?

> One more thingy I'm thinking about: because all this function does is cut-tree reordering, maybe we should move it to `hierarchy.linkage` as optional parameter?

Reply:
Personally, I don't think it belongs in linkage, but I haven't dived deep into the exact purpose of linkage as opposed to dendrogram and where the line is drawn. From my background though, linkage seems like the data model and dendrogram is the view. Keeping these separate would mean that we keep the re-ordering code in dendrogram, as the data model doesn't really care. But again, I could be way off base here. I'll test out the code and get back to you!

Reply:
@ihowell I didn't add the support for checking parameter consistensy yet. Hope to produce it soon, but for now it would rather work as "if You see something weird on the screen then You must've messed up with `leaves_order` "

Reply:
@erheron Floating point distances are showing up again, thanks! But there is one UI issue. You require that the input to `dendrogram` implement the `shape` function, such as a numpy array. This is breaking from the current dendrogram implementation which at the very least allows for a list of linkages, no numpy needed.

Reply:
@ihowell fixed that in recent commit, thanks! Full parameter consistency checking, as well as considering `truncate_mode` (and maybe others) parameters which affect dendrogram is on its way

Reply:
@ihowell, as far as I'm concerned, this is the final code (I wrote more in another thread above). What do You think?